## 01.Importing Libraries

In [1]:
#import libraries
import pandas as pd
import numpy as mp
import os

In [2]:
path = r"C:\Users\praha\OneDrive\Documents\Python\02-2023 Instachart Basket Analysis"

In [3]:
path

'C:\\Users\\praha\\OneDrive\\Documents\\Python\\02-2023 Instachart Basket Analysis'

In [4]:
#importing data set orders.csv
df_ords = pd.read_csv(os.path.join(path, '02. Data', 'Original Data', 'orders.csv'), index_col = False)

In [5]:
df_ords.head()

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2398795        1    prior             2          3                  7   
2    473747        1    prior             3          3                 12   
3   2254736        1    prior             4          4                  7   
4    431534        1    prior             5          4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [6]:
df_ords.shape

(3421083, 7)

Directions 1 & 2: Perform the Wrangling procedures & Find another identifier variable that doesn't need to be included in your analysis as a numeric variable and change it to a suitable format.

In [7]:
#Dropping Columns
df_ords_2 = df_ords.drop(columns = ['eval_set'])

In [8]:
df_ords_2

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0         2539329        1             1          2                  8   
1         2398795        1             2          3                  7   
2          473747        1             3          3                 12   
3         2254736        1             4          4                  7   
4          431534        1             5          4                 15   
...           ...      ...           ...        ...                ...   
3421078   2266710   206209            10          5                 18   
3421079   1854736   206209            11          4                 10   
3421080    626363   206209            12          1                 12   
3421081   2977660   206209            13          1                 12   
3421082    272231   206209            14          6                 14   

         days_since_prior_order  
0                           NaN  
1                          15.0  
2                          21.0  
3                          29.0  
4                          28.0  
...                         ...  
3421078                    29.0  
3421079                    30.0  
3421080                    18.0  
3421081                     7.0  
3421082                    30.0  

[3421083 rows x 6 columns]

In [9]:
df_ords['days_since_prior_order'].value_counts(dropna = False)

30.0    369323
7.0     320608
6.0     240013
4.0     221696
3.0     217005
5.0     214503
NaN     206209
2.0     193206
8.0     181717
1.0     145247
9.0     118188
14.0    100230
10.0     95186
13.0     83214
11.0     80970
12.0     76146
0.0      67755
15.0     66579
16.0     46941
21.0     45470
17.0     39245
20.0     38527
18.0     35881
19.0     34384
22.0     32012
28.0     26777
23.0     23885
27.0     22013
24.0     20712
25.0     19234
29.0     19191
26.0     19016
Name: days_since_prior_order, dtype: int64

In [10]:
df_ords.describe()

order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

In [11]:
#Changing a Variable's Data Type
df_ords['order_id'] = df_ords['order_id'].astype('str') 

In [12]:
df_ords['order_id'].dtype

dtype('O')

In [13]:
df_ords['user_id'] = df_ords['user_id'].astype('str')

In [14]:
df_ords['user_id'].dtype

dtype('O')

In [15]:
df_ords.describe()

order_number     order_dow  order_hour_of_day  days_since_prior_order
count  3.421083e+06  3.421083e+06       3.421083e+06            3.214874e+06
mean   1.715486e+01  2.776219e+00       1.345202e+01            1.111484e+01
std    1.773316e+01  2.046829e+00       4.226088e+00            9.206737e+00
min    1.000000e+00  0.000000e+00       0.000000e+00            0.000000e+00
25%    5.000000e+00  1.000000e+00       1.000000e+01            4.000000e+00
50%    1.100000e+01  3.000000e+00       1.300000e+01            7.000000e+00
75%    2.300000e+01  5.000000e+00       1.600000e+01            1.500000e+01
max    1.000000e+02  6.000000e+00       2.300000e+01            3.000000e+01

Direction 3: Look for a variable in your df_ords dataframe with an unintuitive name and change its name without overwriting the data frame.

In [16]:
#Renaming Columns
df_ords.rename(columns = {'order_dow' : 'orders_day_of_week'}, inplace = True)

In [17]:
df_ords.head()

order_id user_id eval_set  order_number  orders_day_of_week  \
0  2539329       1    prior             1                   2   
1  2398795       1    prior             2                   3   
2   473747       1    prior             3                   3   
3  2254736       1    prior             4                   4   
4   431534       1    prior             5                   4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

Direction 4: Your client wants to know what the busiest hour is for placing orders. Find the frequency of the corresponding variable and share your findings.

In [18]:
#subsetting the highest number of orders (which is 100) to determine the busiest hour
df_busiest_hour = df_ords[df_ords['order_number']==100]

In [19]:
df_busiest_hour

order_id user_id eval_set  order_number  orders_day_of_week  \
3264     3176785     210    train           100                   1   
5122         988     310    train           100                   2   
5243     2956359     313    train           100                   1   
11182    3327840     690    train           100                   4   
12588      90628     786    train           100                   0   
...          ...     ...      ...           ...                 ...   
3408960  1664187  205483    train           100                   2   
3410122  2720227  205543    train           100                   2   
3415733   105274  205878    train           100                   1   
3417312    68079  205972    train           100                   5   
3419501  2411828  206105     test           100                   0   

         order_hour_of_day  days_since_prior_order  
3264                    12                     5.0  
5122                     9                     4.0  
5243                     6                     4.0  
11182                   11                     2.0  
12588                   10                     2.0  
...                    ...                     ...  
3408960                  9                     3.0  
3410122                 18                     8.0  
3415733                 13                     7.0  
3417312                 10                     0.0  
3419501                 11                     1.0  

[1374 rows x 7 columns]

In [20]:
df_busiest_hour.describe()

order_number  orders_day_of_week  order_hour_of_day  \
count        1374.0         1374.000000        1374.000000   
mean          100.0            2.839884          12.950509   
std             0.0            1.912655           4.199906   
min           100.0            0.000000           0.000000   
25%           100.0            1.000000          10.000000   
50%           100.0            3.000000          13.000000   
75%           100.0            4.000000          16.000000   
max           100.0            6.000000          23.000000   

       days_since_prior_order  
count             1374.000000  
mean                 2.723435  
std                  2.479178  
min                  0.000000  
25%                  1.000000  
50%                  2.000000  
75%                  4.000000  
max                 28.000000

In the data above, we can see that mean and median (Q2) in variable 'order_hour_od_day' have the same value, which means the data has normal distribution. Busiest hour = mode = median = mean = 13 (13:00).

Direction 5: Determine the meaning behind a value of 4 in the "department_id" column within the df_prods dataframe using a data dictionary.

In [21]:
#importing data set department.csv
df_dep = pd.read_csv(os.path.join(path, '02. Data', 'Original Data', 'departments.csv'), index_col = False)

In [22]:
df_dep.head()

department_id       1      2       3        4        5              6  \
0    department  frozen  other  bakery  produce  alcohol  international   

           7     8                9  ...            12      13         14  \
0  beverages  pets  dry goods pasta  ...  meat seafood  pantry  breakfast   

             15          16         17      18      19    20       21  
0  canned goods  dairy eggs  household  babies  snacks  deli  missing  

[1 rows x 22 columns]

In [23]:
#transposing data set department.csv
df_dep_t = df_dep.T
df_dep_t.reset_index()

index                0
0   department_id       department
1               1           frozen
2               2            other
3               3           bakery
4               4          produce
5               5          alcohol
6               6    international
7               7        beverages
8               8             pets
9               9  dry goods pasta
10             10             bulk
11             11    personal care
12             12     meat seafood
13             13           pantry
14             14        breakfast
15             15     canned goods
16             16       dairy eggs
17             17        household
18             18           babies
19             19           snacks
20             20             deli
21             21          missing

In [24]:
new_header = df_dep_t.iloc[0]
df_dep_t_new = df_dep_t [1:]
df_dep_t_new.columns = new_header

In [25]:
df_dep_t_new

department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [26]:
data_dict = df_dep_t_new.to_dict('index')

In [27]:
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [28]:
#importing data set products.csv
df_prods = pd.read_csv(os.path.join(path, '02. Data', 'Original Data', 'products.csv'), index_col = False)

In [29]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

In [30]:
#changing a variable's data type
df_prods['product_id'] = df_prods['product_id'].astype('str')

In [31]:
df_prods['product_id'].dtype

dtype('O')

In [32]:
df_prods['aisle_id'] = df_prods['aisle_id'].astype('str')

In [33]:
df_prods['aisle_id'].dtype

dtype('O')

In [34]:
df_prods['department_id'] = df_prods['department_id'].astype('str')

In [35]:
df_prods['department_id'].dtype

dtype('O')

In [36]:
print(data_dict.get('4'))

{'department': 'produce'}


Direction 6: The sales team in your client’s organization wants to know more about breakfast item sales. Create a subset containing only the required information.

In [37]:
#subseting breakfast data (breakfast department = 14)
df_breakfast = df_prods[df_prods['department_id']=='14']

In [38]:
df_breakfast

product_id                                       product_name aisle_id  \
27            28                                  Wheat Chex Cereal      121   
33            34                                                NaN      121   
67            68                            Pancake Mix, Buttermilk      130   
89            90                                       Smorz Cereal      121   
210          211   Gluten Free Organic Cereal Coconut Maple Vanilla      130   
...          ...                                                ...      ...   
49330      49326                            Cereal Variety Fun Pack      121   
49395      49391            Light and Fluffy Buttermilk Pancake Mix      130   
49547      49543                          Chocolate Cheerios Cereal      121   
49637      49633               Shake 'N Pour Buttermilk Pancake Mix      130   
49667      49663  Ultra Protein Power Crunch Peanut Butter N' Ho...       57   

      department_id  prices  
27               14    10.1  
33               14    12.2  
67               14    13.7  
89               14     3.9  
210              14     3.6  
...             ...     ...  
49330            14     9.1  
49395            14     2.0  
49547            14    10.8  
49637            14    14.2  
49667            14    10.2  

[1116 rows x 5 columns]

Direction 7: They’d also like to see details about customers who might be throwing dinner parties. Your task is to find all observations from the entire dataframe that include items from the following departments: alcohol, deli, beverages, and meat/seafood. You’ll need to present this subset to your client.

In [39]:
#subseting dinner data (alcohol = 5, deli = 20 , beverages = 7, meat/seafood = 12)
df_dinner = df_prods.loc[df_prods['department_id'].isin(['5','20','7','12'])]

In [40]:
df_dinner

product_id                                    product_name aisle_id  \
2              3            Robust Golden Unsweetened Oolong Tea       94   
6              7                  Pure Coconut Water With Orange       98   
9             10  Sparkling Orange Juice & Prickly Pear Beverage      115   
10            11                               Peach Mango Juice       31   
16            17                               Rendered Duck Fat       35   
...          ...                                             ...      ...   
49676      49672                          Cafe Mocha K-Cup Packs       26   
49679      49675             Cinnamon Dolce Keurig Brewed K Cups       26   
49680      49676                          Ultra Red Energy Drink       64   
49686      49682                              California Limeade       98   
49688      49684       Vodka, Triple Distilled, Twist of Vanilla      124   

      department_id  prices  
2                 7     4.5  
6                 7     4.4  
9                 7     8.4  
10                7     2.8  
16               12    17.1  
...             ...     ...  
49676             7     6.5  
49679             7    14.0  
49680             7    14.5  
49686             7     4.3  
49688             5     5.3  

[7650 rows x 5 columns]

Direction 8: It’s important that you keep track of total counts in your dataframes. How many rows does the last dataframe you created have?

Dinner data set = 7650 rows

Direction 9: Someone from the data engineers team in Instacart thinks they’ve spotted something strange about the customer with a "user_id" of “1.” Extract all the information you can about this user.

In [41]:
df_userID_1 = df_ords[df_ords['user_id']=='1']

In [42]:
df_userID_1

order_id user_id eval_set  order_number  orders_day_of_week  \
0   2539329       1    prior             1                   2   
1   2398795       1    prior             2                   3   
2    473747       1    prior             3                   3   
3   2254736       1    prior             4                   4   
4    431534       1    prior             5                   4   
5   3367565       1    prior             6                   2   
6    550135       1    prior             7                   1   
7   3108588       1    prior             8                   1   
8   2295261       1    prior             9                   1   
9   2550362       1    prior            10                   4   
10  1187899       1    train            11                   4   

    order_hour_of_day  days_since_prior_order  
0                   8                     NaN  
1                   7                    15.0  
2                  12                    21.0  
3                   7                    29.0  
4                  15                    28.0  
5                   7                    19.0  
6                   9                    20.0  
7                  14                    14.0  
8                  16                     0.0  
9                   8                    30.0  
10                  8                    14.0

Direction 10: You also need to provide some details about this user’s behavior. What basic stats can you provide based on the information you have?

In [43]:
#performing descriptive statistics for User ID 1
df_userID_1.describe()

order_number  orders_day_of_week  order_hour_of_day  \
count     11.000000           11.000000          11.000000   
mean       6.000000            2.636364          10.090909   
std        3.316625            1.286291           3.477198   
min        1.000000            1.000000           7.000000   
25%        3.500000            1.500000           7.500000   
50%        6.000000            3.000000           8.000000   
75%        8.500000            4.000000          13.000000   
max       11.000000            4.000000          16.000000   

       days_since_prior_order  
count               10.000000  
mean                19.000000  
std                  9.030811  
min                  0.000000  
25%                 14.250000  
50%                 19.500000  
75%                 26.250000  
max                 30.000000

Direction 11: Check the organization and structure of your notebook. Be sure to include section headings and code comments.

Checked

Direction 12: Export your df_ords dataframe as “orders_wrangled.csv” in your “Prepared Data” folder.

In [44]:
#export data set 'orders.csv' to 'orders_wrangled.csv'
df_ords.to_csv(os.path.join(path, '02. Data', 'Prepared Data', 'orders_wrangled.csv'))

Direction 13: Export the df_dep_t_new dataframe as “departments_wrangled.csv” in your “Prepared Data” folder so that you have a “.csv” file of your departments data in the correct format.

In [45]:
#export data set 'departments.csv'  to 'departments_wrangled.csv'
df_dep_t_new.to_csv(os.path.join(path, '02. Data', 'Prepared Data', 'departments_wrangled.csv'))